movement/analysis/filtering.py

```diff
@@ -0,0 +1,141 @@
+    limit : int | None
+        Maximum number of consecutive NaNs to interpolate over.
+        `None` indicates no limit, and is the default value.
+    max_gap : TODO: Clarify the difference between `limit` & `max_gap`
+        The largest gap of consecutive NaNs that will be
+        interpolated over. The default value is `None`.
+    inplace: bool
+        If true, updates the provided DataSet in place and returns
+        `None`.
+
+    Returns
+    -------
+    ds_thresholded : xr.DataArray
+        The provided dataset (ds), where NaN values have been
+        interpolated over using the parameters provided.
+    """
+    # TODO: This method interpolates over confidence values as well.
+    #  -> Figure out whether this is the desired default behavior.
+    ds_interpolated = ds.interpolate_na(
+        dim="time", method=method, limit=limit, max_gap=max_gap
+    )
+
+    # Logging
+    if "log" not in ds_interpolated.attrs.keys():
+        ds_interpolated.attrs["log"] = []
+
+    log_entry = {
+        "operation": "interp_pose",
+        "method": method,
+        "limit": limit,
+        "max_gap": max_gap,
+        "inplace": inplace,
+        "datetime": str(datetime.now()),
+    }
+    ds_interpolated.attrs["log"].append(log_entry)
+
+    if inplace:
+        ds["pose_tracks"] = ds_interpolated["pose_tracks"]
+        ds["confidence"] = ds_interpolated["confidence"]
+        ds.attrs["log"] = ds_interpolated.attrs["log"]
+        return None
+    else:
+        return ds_interpolated
+
+
+def filter_confidence(
+    ds: xr.Dataset,
+    threshold: float = 0.6,
+    inplace: bool = False,
+    interp: bool = False,
+) -> Union[xr.Dataset, None]:
+    """
+    Drops all datapoints where the associated confidence value
+    falls below a user-defined threshold.
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        Dataset containing pose tracks, confidence scores, and metadata.
+    threshold : float
+        The confidence threshold below which datapoints are filtered.
+        A default value of `0.6` is used.
+    inplace : bool
+        If true, updates the provided DataSet in place and returns
+        `None`.
+    interp : bool
+        If true, NaNs are interpolated over using `interp_pose` with
+        default parameters.
+
+    Returns
+    -------
+    ds_thresholded : xarray.Dataset
+        The provided dataset (ds), where datapoints with a confidence
+        value below the user-defined threshold have been converted
+        to NaNs
+    """
+
+    ds_thresholded = ds.where(ds.confidence >= threshold)
+
+    # Diagnostics
+    print("\nDatapoints Filtered:\n")
+    for kp in ds.keypoints.values:
+        n_nans = np.count_nonzero(
+            np.isnan(ds_thresholded.confidence.sel(keypoints=f"{kp}").values)
+        )
+        n_points = ds.time.values.shape[0]
+        prop_nans = round((n_nans / n_points) * 100, 2)
+        print(f"{kp}: {n_nans}/{n_points} ({prop_nans}%)")
```

Comment:
This calculations don't work properly. For examply, on the "SLEAP_three-mice_Aeon" dataset I get:
```
Datapoints Filtered:

centroid: 1803/601 (300.0%)
```

There are several problems here:
- this dataset has NaN confidence values, in such cases we should not filter (so drop a value only if confidence is not None and is below the theshold)
- the 300% is obviously wrong, it counts all 3 centroids (from all 3 individuals). Perhaps it's better split the report by individual and by keypoint (instead of by keypoint only).

Reply:
Thanks for the heads up here! The diagnostics block should now be fixed!